File: app/src/main/java/net/bible/service/history/HistoryTraversal.java
/*
 * Copyright (c) 2018 Martin Denham, Tuomas Airaksinen and the And Bible contributors.
 *
 * This file is part of And Bible (http://github.com/AndBible/and-bible).
 *
 * And Bible is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * And Bible is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with And Bible.
 * If not, see http://www.gnu.org/licenses/.
 *
 */

package net.bible.service.history;

import android.util.Log;

import net.bible.android.control.event.ABEventBus;
import net.bible.android.control.event.passage.BeforeCurrentPageChangeEvent;


/**
 * @author Martin Denham [mjdenham at gmail dot com]
 */
public class HistoryTraversal {

	private boolean integrateWithHistoryManager;
	
	private long lastBackNavTime;

	private final HistoryManager historyManager;
	
	private static long MIN_BACK_SEPERATION_MILLIS = 500;

	private static String TAG = "HistoryTraversal";

	public HistoryTraversal(HistoryManager historyManager, boolean integrateWithHistoryManager) {
		this.historyManager = historyManager;
		this.integrateWithHistoryManager = integrateWithHistoryManager;
	}

	/**
     * about to change activity so tell the HistoryManager so it can register the old activity in its list
     */
	public void beforeStartActivity() {
		if (integrateWithHistoryManager) {
			ABEventBus.getDefault().post(new BeforeCurrentPageChangeEvent());
		}
	}

	public boolean goBack() {
		long prevBackNavTime = lastBackNavTime;
		lastBackNavTime = System.currentTimeMillis();
		if (lastBackNavTime-prevBackNavTime<MIN_BACK_SEPERATION_MILLIS) {
			// swallow back key if it seems like a phantom repeat to prevent history item jumping
			return true;
		} else if (integrateWithHistoryManager && historyManager.canGoBack()) {
			Log.d(TAG, "Go back");
			historyManager.goBack();
			return true;
		} else {
			return false;
		}
	}

	public boolean isIntegrateWithHistoryManager() {
		return integrateWithHistoryManager;
	}

	public void setIntegrateWithHistoryManager(boolean integrateWithHistoryManager) {
		this.integrateWithHistoryManager = integrateWithHistoryManager;
	}
}

File: app/src/main/java/net/bible/service/history/HistoryTraversal.kt
/*
 * Copyright (c) 2020-2022 Martin Denham, Tuomas Airaksinen and the AndBible contributors.
 *
 * This file is part of AndBible: Bible Study (http://github.com/AndBible/and-bible).
 *
 * AndBible is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * AndBible is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with AndBible.
 * If not, see http://www.gnu.org/licenses/.
 */

package net.bible.service.history

import android.util.Log

import net.bible.android.control.event.ABEventBus
import net.bible.android.control.event.passage.BeforeCurrentPageChangeEvent


/**
 * @author Martin Denham [mjdenham at gmail dot com]
 */
class HistoryTraversal(val historyManager: HistoryManager, var isIntegrateWithHistoryManager: Boolean) {

    /**
     * about to change activity so tell the HistoryManager so it can register the old activity in its list
     */
    fun beforeStartActivity() {
        if (isIntegrateWithHistoryManager) {
            ABEventBus.post(BeforeCurrentPageChangeEvent())
        }
    }

    fun goBack(): Boolean {
        if (isIntegrateWithHistoryManager && historyManager.canGoBack()) {
            Log.i(TAG, "Go back")
            historyManager.goBack()
            return true
        } else {
            return false
        }
    }

    companion object {

        private val MIN_BACK_SEPERATION_MILLIS: Long = 500

        private val TAG = "HistoryTraversal"
    }
}
